quantization/src/encoded_vectors_pq.rs

```diff
@@ -243,6 +323,69 @@ impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
         Ok(result)
     }
 
+    #[allow(unused)]
+    pub fn find_centroids_rayon<'a>(
+        data: impl IndexedParallelIterator<Item = &'a [f32]> + Clone,
+        vector_division: &[Range<usize>],
+        vector_parameters: &VectorParameters,
+        centroids_count: usize,
+        max_kmeans_threads: usize,
+    ) -> Result<Vec<Vec<f32>>, EncodingError> {
+        let sample_size = KMEANS_SAMPLE_SIZE.min(vector_parameters.count);
+        let mut result = vec![vec![]; centroids_count];
+
+        // if there are not enough vectors, set centroids as point positions
+        if vector_parameters.count <= centroids_count {
+            data.map(|vector_data| vector_data.to_vec())
+                .collect_into_vec(&mut result);
+
+            for r in result
+                .iter_mut()
+                .take(centroids_count)
+                .skip(vector_parameters.count)
+            {
+                *r = vec![0.0; vector_parameters.dim];
+            }
+            return Ok(result);
+        }
+
+        // generate random subset of data
+        let permutor = permutation_iterator::Permutor::new(vector_parameters.count as u64);
+        let mut selected_vectors: Vec<usize> =
+            permutor.map(|i| i as usize).take(sample_size).collect();
+        selected_vectors.sort_unstable();
+
+        for range in vector_division.iter() {
+            let mut data_subset = Vec::with_capacity(sample_size * range.len());
+
+            // TODO: not sure if this impl is the same as before
+            let mut par_iter = data
+                .clone()
+                .enumerate()
+                .filter(|(vector_index, _vector_data)| {
+                    // TODO: what we did before with selected_index is probably more efficient?
+                    selected_vectors.binary_search(vector_index).is_ok()
+                })
+                .take_any(sample_size)
+                .flat_map(|(_vector_index, vector_data)| &vector_data[range.clone()]);
+            data_subset.par_extend(par_iter);
+
+            let centroids = kmeans(
```

Comment:
This internally spawns a new rayon threadpool, which isn't required if this uses a threadpool as well.